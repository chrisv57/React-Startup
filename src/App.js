import * as React from 'react'

// const welcome= {
//   greeting: 'Hey',
//   title: 'Mars and Reaact',
// };

function getTitle(title) {
  return title;
}
const list = [
  {
    title: 'React',
    url: 'https://reactjs/org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
  },
]
const App = () =>
  (
    <div>
      <h1>Hey {getTitle('React')}</h1>
    
    <Search />
    <hr />
    
    <List />
   {/* Two instances of List component */}
    <List/>
    </div>
    
  );

const Search = () =>
  (
    <div>
      <label htmlFor='search'>Search: </label>
      <input id ="search" type="text"/>
    </div>
  );

//definiton of list component
const List = () =>{
  //peform a task in between
  return (
    <div>
      {/* list rendeirng*/}
      <ul>
        {list.map((item) => {
           return(
             <li key = {item.objectID}>
               <span>
                 <a href={item.url}>{item.title} </a>
               </span>
               <span>{item.author}</span>
               <span>{item.num_comments}</span>
               <span>{item.points}</span>
               <span>{null}</span>
             </li>
           );
        })}
      </ul>
    </div>
  )
}

export default App;