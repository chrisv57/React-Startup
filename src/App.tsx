import * as React from "react";
import axios from "axios";

import './App.css';
import { ReactComponent as Cross } from "./cross.svg";

const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

const useSemiPersistentState = (
  key: string,
  initalState: string
): [string, (newValue: string) => void] => {
  const isMounted = React.useRef(false);
  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initalState
  );
  React.useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = true;
    } else {
      console.log("A");
      localStorage.setItem(key, value);
    }
  }, [value, key]);
  return [value, setValue];
};

type Story = {
  objectID: string;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};
type Stories = Array<Story>;

type StoriesState = {
  data: Stories;
  isLoading: boolean;
  isError: boolean;
};

interface StoriesFetchInitAction{
  type: 'STORIES_FETCH_INIT';
}

interface StoriesFetchSuccessAction{
  type: 'STORIES_FETCH_SUCCESS';
  payload: Stories;
}

interface StoriesFetchFailureAction{
  type: 'STORIES_FETCH_FAILURE';
}

interface StoriesFetchRemoveAction{
  type: 'REMOVE_STORY';
  payload: Story;
}
type StoriesAction = 
  | StoriesFetchInitAction
  | StoriesFetchSuccessAction
  | StoriesFetchFailureAction
  | StoriesFetchRemoveAction;



const storiesReducer = (state: StoriesState, action: StoriesAction) => {
  switch (action.type) {
    case "STORIES_FETCH_INIT":
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case "STORIES_FETCH_SUCCESS":
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case "STORIES_FETCH_FAILURE":
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case "REMOVE_STORY":
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

const getSumComments = (stories) => {
  console.log('C');

  return stories.data.reduce(
    (result, value) => result + value.num_comments,
    0
  );
};



const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");

  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);

  //Merging states using useReducer hook
  const [stories, dispatchStories] = React.useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });

  const handleFetchStories = React.useCallback(async () => {
    dispatchStories({ type: "STORIES_FETCH_INIT" });

    try {
      const result = await axios.get(url);
      dispatchStories({
        type: "STORIES_FETCH_SUCCESS",
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: "STORIES_FETCH_FAILURE" });
    }
  }, [url]);
  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = React.useCallback((item: Story) => {
    dispatchStories({
      type: "REMOVE_STORY",
      payload: item,
    });
  }, []);

  // const handleSearch = (event) => {
  //   setSearchTerm(event.target.value);
  // }

  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (event: React.ChangeEvent<HTMLFormElement>) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  console.log('B:APP');
  const sumComments = React.useMemo(() => getSumComments(stories), [
    stories,
  ]);
  return (
    <div className="container">
      <h1 className="headline-primary">Hello React</h1>
      <h1>Number of Comments {sumComments} </h1>

      {/* <Search /> */}
      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />
      {/* In JavaScript, a true && 'Hello World' always evaluates to ‘Hello World’. A false && 'Hello World'
always evaluates to false. */}
      {stories.isError && <p>Something went wrong....</p>}
      {stories.isLoading ? (
        <p>Loading....</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </div>
  );
};

type SearchFormProps = {
  searchTerm: string;
  onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onSearchSubmit: (event: React.ChangeEvent<HTMLFormElement>) => void;
}

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit,
}: SearchFormProps) => (
  <form onSubmit={onSearchSubmit} className="search-form">

    <InputWithLabel
      id="search"
      value={searchTerm}
      isFocused
      onInputChange={onSearchInput}
    >
      <strong>Search:</strong>
    </InputWithLabel>
    <button type="submit" disabled={!searchTerm}
      className="button button_large"
    >
      Submit
    </button>

  </form>

);


type InputWithLabelProps = {
  id: string;
  value: string;
  type?: string;
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) =>void;
  isFocused?: boolean;
  children: React.ReactNode;
};

const InputWithLabel = ({
  id,
  value,
  type = "text",
  onInputChange,
  isFocused,
  children,
}: InputWithLabelProps) => {
  const inputRef = React.useRef<HTMLInputElement>(null!);
  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);
  return (
    <>
      <label
        htmlFor={id}
        className="label"
      >
        {children}
      </label>
      &nbsp;
      <input
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
        className="input"
      />
    </>
  );
};

type ListProps = {
  list: Stories;
  onRemoveItem: (item: Story) => void;
};

//USing Spread and Rest Operators
const List = ({ list, onRemoveItem }: ListProps) =>
(
  <ul>
    {list.map((item) => (
      <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
    ))}
  </ul>
);



type itemProps = {
  item: Story;
  onRemoveItem: (item: Story) => void;
};
// The item is of type Story; the onRemoveItem function takes an item of type Story as an argument
// and returns nothing.
const Item = ({ item, onRemoveItem, }: itemProps) => {
  return (
    <li className="item">
      <span style={{ width: '40%' }}>
        <a href={item.url}>{item.title}</a>
      </span>
      <span style={{ width: '30%' }}>{item.author}</span>
      <span style={{ width: '10%' }}>{item.num_comments}</span>
      <span style={{ width: '10%' }}>{item.points}</span>
      <span style={{ width: '10%' }}>
        <button
          type="button"
          onClick={() => onRemoveItem(item)}
          className="button button_small"
        >
          <Cross height="18px" width="18px" />
        </button>
      </span>
    </li>
  );
};
export default App;

export {storiesReducer, SearchForm, InputWithLabel, List, Item};
