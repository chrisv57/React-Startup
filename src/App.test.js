import { render, screen } from '@testing-library/react';
import App from './App';

// test('renders learn react link', () => {
//   render(<App />);
//   const linkElement = screen.getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });


// The “describe” block is our test suite, and the “test” blocks are our test cases.

describe('Something true and false',()=>{
  test('true to be true',()=>{
    expect(true).toBeTruthy();
  });

  test('false to be false',()=>{
  expect(false).toBeFalsy();
  });
});


//Test blocks can also be written as "it".
// describe('App Component',()=>{
//   it('removes an item when clicking the dismiss button',()=>{

//   });

//   it('request initial stoires from the API',()=>{

//   });
// });
