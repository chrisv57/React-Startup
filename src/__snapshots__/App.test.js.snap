// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm render snapshot 1`] = `
<form
  class="search-form"
>
  <label
    class="label"
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  Â 
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="button button_large"
    type="submit"
  >
    Submit
  </button>
</form>
`;
